loclist.kak

```diff
@@ -0,0 +1,147 @@
+            evaluate-commands -draft -save-regs / %{
+                set-register / %opt{loclist_location_format}
+                execute-keys <semicolon>xs<ret>
+                set-register a "%reg{1}"
+                set-register b "%reg{2}"
+                set-register c "%reg{3}"
+                try %{
+                    # Is it an absolute path?
+                    execute-keys s\A/.*<ret>
+                } catch %{
+                    set-register a "%opt{loclist_root}%reg{a}"
+                }
+            }
+            set-option buffer grep_current_line %val{cursor_line}
+            evaluate-commands -try-client %opt{jumpclient} -verbatim -- edit -existing -- %reg{a} %reg{b} %reg{c}
+            try %{ focus %opt{jumpclient} }
+        }
+    }
+}
+
+define-command loclist-next-location -docstring %{
+    loclist-next-location
+    Jump to next location listed in the current location list buffer, usually one of
+    *diagnostics* *goto* *grep* *implementations* *lint-output* *make* *references* *symbols*
+
+    %opt{loclist_buffer} determines the buffer current location list buffer.
+    %opt{loclist_location_format} determines matching locations.
+} -buffer-completion %{
+    evaluate-commands -try-client %opt{jumpclient} -save-regs / %{
+        buffer %opt{loclist_buffer}
+        set-register / %opt{loclist_location_format}
+        execute-keys ge %opt{grep_current_line}g<a-l> /<ret>
+        loclist-jump
+    }
+    try %{
+        evaluate-commands -client %opt{toolsclient} %{
+            buffer %opt{loclist_buffer}
+            execute-keys gg %opt{grep_current_line}g
+        }
+    }
+}
+
+define-command loclist-previous-location -docstring %{
+    loclist-previous-location
+    Jump to previous location listed in the current location list buffer, usually one of
+    *diagnostics* *goto* *grep* *implementations* *lint-output* *make* *references* *symbols*
+
+    %opt{loclist_buffer} determines the buffer current location list buffer.
+    %opt{loclist_location_format} determines matching locations.
+} -buffer-completion %{
+    evaluate-commands -try-client %opt{jumpclient} -save-regs / %{
+        buffer %opt{loclist_buffer}
+        set-register / %opt{loclist_location_format}
+        execute-keys ge %opt{grep_current_line}g<a-h> <a-/><ret>
+        loclist-jump
+    }
+    try %{
+        evaluate-commands -client %opt{toolsclient} %{
+            buffer %opt{loclist_buffer}
+            execute-keys gg %opt{grep_current_line}g
+        }
+    }
+}
+
+define-command -override loclist-stack-push -docstring "record location list buffer" %{
+    evaluate-commands %sh{
+        eval set -- $kak_quoted_opt_loclist_stack
+        if printf '%s\n' "$@" | grep -Fxq -- "$kak_bufname"; then {
+            exit
+        } fi
+        newbuf=$kak_bufname-$#
+        echo "try %{ delete-buffer! $newbuf }"
+        echo "rename-buffer $newbuf"
+        echo "set-option -add global loclist_stack %val{bufname}"
```

Comment:
I need to look into whether this unnecessarily pollutes the buffer list.
Think `buffer-next`/`buffer-previous` (which I never use).
If it turns out to be problematic, we can implement `arrange-buffers -end` to move all but the top location buffer to the end.
The reason why I think this might work is because I never switch to the non-top location buffers except through `locations-pop` (which should move the new top buffer to the front again).

Reply:
We can also consider splitting the buffer stack stuff out from this. But its very useful like this as well

Reply:
Yeah in future we can have a generic buffer-stack module because the git-stack has essentially the same code.
I don't anticipate many changes to it but it's still nicer.

One thing we might want to do in future is to limit the buffer stack size. I have a vague memory of this being tricky (or at least not as simple as the current iml)